torch/csrc/jit/codegen/cuda/index_compute.cpp

```diff
@@ -1939,6 +1939,186 @@ std::pair<std::vector<kir::Val*>, bool> Index::getConsumerRootPredIndices(
+    if (extent != nullptr) {
+      reference_halo_extent_map[gpu_lower->lowerValue(ref_id)
+                                    ->as<kir::IterDomain>()] =
+          gpu_lower->lowerValue(extent);
+    }
+  }
+
+  // Index into the reference tensor
+  auto ref_indexing = getReferenceIndexing(
+      loops,
+      reference_domain,
+      ref_id_to_ind_map,
+      {},
+      reference_halo_extent_map);
+
+  // // Index into consumer using reference indexing
+  // auto consumer_indexing = ref_compute.updateIndexCompute(
+  //     consumer_tv->domain(),
+  //     ref_2_consumer,
+  //     root_contiguity,
+  //     reference_halo_extent_map);
+
+  // Indices should now be mapped onto IterDomains in consumer, so just grab
+  // and use them.
+
+  // If we are generating a predicate for initialization, we should use
+  // rfactor instead of root_dom. If we are generating a predicate for
+  // actual reduction expr, reduction axes should have their indices
+  // mapped to non-zero symbolic vals.
+
+  // Map consumer tensor to reference
+
+  std::unordered_map<
+      torch::jit::fuser::cuda::IterDomain*,
+      torch::jit::fuser::cuda::IterDomain*>
+      consumer_2_ref;
+
+  for (auto entry : ref_2_consumer) {
+    auto ref_id = entry.first;
+    auto c_id = entry.second;
+    consumer_2_ref[c_id] = ref_id;
+  }
+
+  bool buffer_init = false;
+  for (auto consumer_id : consumer_tv->domain()->domain()) {
+    if (consumer_id->isReduction()) {
+      if (consumer_2_ref.count(consumer_id)) {
+        auto kir_ref_id = gpu_lower->lowerValue(consumer_2_ref.at(consumer_id))
+                              ->as<kir::IterDomain>();
+        auto ref_id_idx_it = ref_indexing.indexMap().find(kir_ref_id);
+        if (ref_id_idx_it != ref_indexing.indexMap().end() &&
+            !ref_id_idx_it->second->isZeroInt()) {
+          // If there's an index value for a reduction dimension of consumer
+          // we must not be initializing consumer_tv
+          buffer_init = false;
+          break;
+        }
+      }
+      buffer_init = true;
+    }
+  }
+
+  // If we are initializing a reduction buffer and the tensor has a
+  // rfactor root, the predicate should be based on the rfactor root.
+  const auto root_domain = reference_domain->getRootDomain();
+
+  std::vector<kir::Val*> root_inds(root_domain.size(), ir_builder.zeroVal());
+
+  ReferencePredicate rp;
+  rp.mapped_to_rfactor = buffer_init;
+  rp.reference_domain = reference_domain;
+
+  for (size_t i = 0; i < root_domain.size(); i++) {
+    if (root_domain[i]->isBroadcast() ||
+        gpu_lower->trivialReductionInfo().isDerived(root_domain[i])) {
+      continue;
+    }
+    auto kir_root_id =
+        gpu_lower->lowerValue(root_domain[i])->as<kir::IterDomain>();
+    const auto it = ref_indexing.indexMap().find(kir_root_id);
+    if (it != ref_indexing.indexMap().end()) {
+      root_inds[i] = it->second;
+      if (ref_indexing.zeroMergedIn().count(kir_root_id)) {
```

Comment:
I'm confused here. When creating `ref_indexing`, it doesn't seem to look at the allocation point of the consumer tensor, so why can there be zero-merged-in domains?